From 045f74dad4079f7eab23b4f8070d04d606eb88bc Mon Sep 17 00:00:00 2001
From: shamim ali <ashamim@ami.com>
Date: Thu, 12 Jun 2025 06:43:24 -0400
Subject: [PATCH] Add new certificate data fields support

This patch will add new dbus properties such as:
- SerialNumber
- PublicKey
- SignatureAlgorithm
- CertificateVersion

Signed-off-by: shamim ali <ashamim@ami.com>
---
 certificate.cpp | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/certificate.cpp b/certificate.cpp
index 34370b6..70e1e5f 100644
--- a/certificate.cpp
+++ b/certificate.cpp
@@ -32,6 +32,9 @@
 #include <map>
 #include <utility>
 #include <vector>
+#include <openssl/x509.h>
+#include <openssl/bn.h>
+
 
 namespace phosphor::certs
 {
@@ -584,6 +587,29 @@ void Certificate::populateProperties(X509& cert)
     ASN1_TIME* notBefore = X509_get_notBefore(&cert);
     ASN1_TIME_diff(&days, &secs, epoch.get(), notBefore);
     validNotBefore((days * dayToSeconds) + secs);
+
+    EVP_PKEY* pkey = X509_get_pubkey(&cert);
+    uint16_t keySize = EVP_PKEY_bits(pkey);  // returns key size in bits, e.g., 2048
+    EVP_PKEY_free(pkey);
+    publicKey(keySize);
+
+    ASN1_INTEGER* serial = X509_get_serialNumber(&cert);
+    BIGNUM* bn = ASN1_INTEGER_to_BN(serial, NULL);
+    char* hex = BN_bn2hex(bn);
+    std::string serialStr(hex);
+    serialNumber(serialStr);
+    BN_free(bn);
+    OPENSSL_free(hex);
+
+    int sig_nid = X509_get_signature_nid(&cert);
+    const char* sigAlgo = OBJ_nid2ln(sig_nid);  // long name (e.g., "sha256WithRSAEncryption")
+    std::string signatureAlgo(sigAlgo);
+    signatureAlgorithm(signatureAlgo);
+
+    long version = X509_get_version(&cert);  // 0 = v1, 1 = v2, 2 = v3
+    std::string versionStr = std::to_string(version + 1);  // Convert to human-readable v1/v2/v3
+    certificateVersion(versionStr);
+
 }
 
 void Certificate::checkAndAppendPrivateKey(const std::string& filePath)
-- 
2.34.1

